### DISCUSSION TITLE
OData Metadata Parser

### DISCUSSION THREAD

**Discussion by ghiscoding** [ghiscoding/slickgrid-universal#578] [Ideas]
Hello @der-david, hope you don't mind that I'm reaching out to you this way.

So I was looking at opened forks of my lib and came across yours by luck and I found some very very interesting code that you started working on for an OData Metadata Parser. It looks like you stopped working on it but I was wondering if you would be interested in advancing it and possibly making this as a contribution back into the lib? I think you had some problem to get it running but you were really so close (I changed less than 20 lines of code) and I kind of got it working (at least a portion of it) as can be seen in the print screen below (it reads the metadata and creates a grid with 2 columns), we don't see anything in the grid because CORS is blocking me from executing the final OData query since I was trying with an OData sample from Microsoft (which is not part of my domain hence why I'm getting the CORS issue, I even tried installing a CORS Chrome extension but that doesn't help for executing the last query)
[https://services.odata.org/TripPinRESTierService/(S(dh01mgakplqunvumgf2rs2xn))/$metadata#Airlines](https://services.odata.org/TripPinRESTierService/(S(dh01mgakplqunvumgf2rs2xn))/$metadata#Airlines)
 
Do I understand correctly that this service that you started would basically just read the OData metadata and create a grid totally dynamically on the fly without having to create any grid column definitions at all? If I understood correctly then that is huge, it's a big game changer for many developers. So anyway, I don't have access to any OData server anymore but I find your work so very interesting which is why I'm attempting to reach out to you through GitHub (hopefully you can read this and you won't be mad at me for peaking at your code). 

If you would be interested in pursuing this project, I'm sure there would be many developers, for example @jr01, interested into this potential awesome OData optional package. 

Have nice day 🌞 
@ghiscoding 

![image](https://user-images.githubusercontent.com/643976/146126382-4fd721ca-bc7c-43e2-8475-885bc750e0b2.png)

**Comment by ghiscoding** [comment]
and 2 hours later... it works! Well not all of it but the concept is working.
So I found this [example](https://github.com/janhommes/o.js/blob/master/example/node.js) that `o.js` is using, used the same endpoint as them, modified a few things with the OData `getData` and I'm getting something. By the way, nice little lib, I didn't know about it.

Dynamic grid creation with OData...ohhh this is starting to be really interesting :)

A few things to note from demo below:
- the count is not working, at least it doesn't show in result, so I made it fixed to 10 for now
   - existing [issue](https://github.com/janhommes/o.js/issues/105) in `o.js`, doesn't seem feasible in same query :(
- the "AddressInfo" is an object, so I guess the parser would have to split that column generation into multiple columns maybe!?

![2021-12-15_02-30-26](https://user-images.githubusercontent.com/643976/146142779-10e5aafc-1765-4234-813b-e2a13b554c80.gif)

**Comment by jr01** [comment]
I think this is a great idea @ghiscoding, especially to get a new project going fast.

In my current project we have disabled the metadata endpoint for security (obscurity...) reasons and also we have a lot of customization in each grid, so I won't be using this.

Auto-splitting `AddressInfo` into columns is an interesting thought. If it is a 'complex type' then filtering and sorting the individual properties on the backend might be a challenge. Another approach is to just show the object with `JSON.stringify`

If `AddressInfo` is a navigation (like a 1:N related entity) however, then I have some code that I intended to contribute back for some time now. I'll paste that into a new discussion...

**Reply by ghiscoding** [reply]
Well for the auto-splitting, I was actually thinking that perhaps Column Header Grouping (like [Example 8](https://ghiscoding.github.io/slickgrid-universal/#/example08) might actually be a better option and what is underneath (AddressInfo) would go into that group (as long as it's not object of objects of objects...)

I find the project interesting because I can probably do the same with GraphQL (which I use a bit more) and they probably have metadata exposed as well, but yeah I didn't think about the security side of it. It might be good for Admin dashboard or something like that but it's probably an all or nothing for the metadata endpoint. Another option might be to save only the list of column names (and maybe their type) into a stringified JSON string and then run it through that parser, this would still allow for some dynamic grid creation. 

Anyway it's an interesting project, I was just curious and gave it a try, the concept is working and @der-david did most of the work on the OData metadata parser.

**Comment by der-david** [comment]
@ghiscoding @jr01 sorry for the delay, happy new year! I've been busy with other stuff and couldn't really take care of the repos and github.

Yes, the idea is to dynamically generate an editable SlickGrid based on the Metadata of an arbitrary OData service and provide the respective I/O-components (column titles, input fields etc.) for the different datatypes. I think the branch is wayyy behind my local developments, but then again even those have been neglected for a few months now. I'll see if I can clean up the code and push it so you can have a better overview.

IIRC, I was fiddling around in many places to make my changes work consistently (sorting, pagination etc.) and to prevent unsaved changes from being discarded upon certain actions. So I may have to split the feature(s) into several branches to allow for a clean PR merge procedure.

**Reply by ghiscoding** [reply]
@der-david I can send you the changes I made (which by the way, were from your branch, so it's also not up to date) to have it work the way it is shown in the last animated gif, I'm not intending to do more changes since I'm not using OData anymore but would happily send the change and/or receive a Pull Request. Cheers :)
